pallets/transfer-allowlist/src/lib.rs

```diff
@@ -0,0 +1,481 @@
+		/// Increments number of allowances present for a sending account/currency set.
+		/// If no allowances set, an entry with 1 added, if entry already present, it is then incremented.
+		pub fn increment_or_create_allowance_count(
+			account_id: &T::AccountId,
+			currency_id: &T::CurrencyId,
+		) -> DispatchResult {
+			// not using try_mutate here as we're not sure if key exits, and we're already doing a some value check on result of exists query check
+			match <AccountCurrencyTransferRestriction<T>>::get(account_id, currency_id) {
+				Some(allowance_count) if allowance_count > 0 => {
+					let new_allowance_count = allowance_count
+						.checked_add(1)
+						.ok_or(Error::<T>::AllowanceCountOverflow)?;
+					<AccountCurrencyTransferRestriction<T>>::insert(
+						account_id,
+						currency_id,
+						new_allowance_count,
+					);
+					Ok(())
+				}
+				Some(_) => Err(DispatchError::Other("Impossible allowance count")),
+				_ => {
+					<AccountCurrencyTransferRestriction<T>>::insert(account_id, currency_id, 1);
+					Ok(())
+				}
+			}
+		}
+
+		/// Decrements the number of allowances tracked for a sending account/currency set.
+		/// If the allowance count is currently 1, then it removes the entry
+		/// If greater than 1, then decremented.
+		/// If no entry present, NoAllowancesSet error returned.
+		pub fn decrement_or_remove_allowance_count(
+			account_id: &T::AccountId,
+			currency_id: &T::CurrencyId,
+		) -> DispatchResult {
+			// not using try_mutate here as we're not sure if key exits, and we're already doing a some value check on result of exists query check
+			match <AccountCurrencyTransferRestriction<T>>::get(account_id, currency_id) {
+				Some(allowance_count) if allowance_count <= 1 => {
+					<AccountCurrencyTransferRestriction<T>>::remove(account_id, currency_id);
+					Ok(())
+				}
+				Some(allowance_count) => {
+					// check in this case should not ever be needed
+					let new_allowance_count = allowance_count.ensure_sub(1)?;
+					<AccountCurrencyTransferRestriction<T>>::insert(
+						account_id,
+						currency_id,
+						new_allowance_count,
+					);
+					Ok(())
+				}
+				_ => Err(DispatchError::from(Error::<T>::NoAllowancesSet)),
+			}
+		}
+	}
+
+	impl<T: Config> TransferAllowance<T::AccountId> for Pallet<T> {
+		type CurrencyId = T::CurrencyId;
+		type Location = Location;
+
+		/// This checks to see if a transfer from an account and currency should be allowed to a given location.
+		/// If there are no allowances defined for the sending account and currency, then the transfer is allowed.
+		/// If there is an allowance for the sending account and currency,
+		/// but the destination does not have an allowance added then the transfer is not allowed.
+		/// If there is an allowance for the sending account and currency,
+		/// and there's an allowance present:
+		/// then we check whether the current block is between the `allowed_at` and `blocked_at` blocks in the allowance.
+		fn allowance(
```

Comment:
Do you have an example of how this will be used by a consumer pallet in a runtime?

Reply:
Iirc we were planning on using a `PreConditions` impl that would call the allowance trait fn on the pallet to filter out whether the request should be allowed or not.